shared-module/tilepalettemapper/TilePaletteMapper.c

```diff
@@ -0,0 +1,72 @@
+// This file is part of the CircuitPython project: https://circuitpython.org
+//
+// SPDX-FileCopyrightText: Copyright (c) 2025 Tim Cocks for Adafruit Industries
+//
+// SPDX-License-Identifier: MIT
+#include <stdlib.h>
+#include "py/runtime.h"
+#include "shared-bindings/tilepalettemapper/TilePaletteMapper.h"
+#include "shared-bindings/displayio/Palette.h"
+
+void common_hal_tilepalettemapper_tilepalettemapper_construct(tilepalettemapper_tilepalettemapper_t *self,
+    mp_obj_t palette, uint16_t width, uint16_t height) {
+
+    self->palette = palette;
+    self->width_in_tiles = width;
+    self->height_in_tiles = height;
+    self->needs_refresh = false;
+    int mappings_len = width * height;
+    self->tile_mappings = (uint16_t **)m_malloc(mappings_len * sizeof(uint16_t *));
+    uint32_t palette_len = common_hal_displayio_palette_get_len(self->palette);
+    for (int i = 0; i < mappings_len; i++) {
+        self->tile_mappings[i] = (uint16_t *)m_malloc(palette_len * sizeof(uint16_t));
+        for (uint16_t j = 0; j < palette_len; j++) {
+            self->tile_mappings[i][j] = j;
+        }
+    }
+}
+
+uint16_t common_hal_tilepalettemapper_tilepalettemapper_get_width(tilepalettemapper_tilepalettemapper_t *self) {
+    return self->width_in_tiles;
+}
+
+uint16_t common_hal_tilepalettemapper_tilepalettemapper_get_height(tilepalettemapper_tilepalettemapper_t *self) {
+    return self->height_in_tiles;
+}
+
+mp_obj_t common_hal_tilepalettemapper_tilepalettemapper_get_palette(tilepalettemapper_tilepalettemapper_t *self) {
+    return self->palette;
+}
+
+mp_obj_t common_hal_tilepalettemapper_tilepalettemapper_get_mapping(tilepalettemapper_tilepalettemapper_t *self, uint16_t x, uint16_t y) {
+    mp_obj_list_t *list = MP_OBJ_TO_PTR(mp_obj_new_list(0, NULL));
```

Comment:
Return a `tuple` instead of a list because `tuples` aren't mutable. I'm worried that mutating the list will be expected to change behavior.

Reply:
updated in latest commit to run a tuple